File: aspnetcore/signalr/api-design/sample/Samples.cs
#region ParameterBasedOldVersion
public async Task<string> GetTotalLength(string param1)
{
    return param1.Length;
}
#endregion

#region ParameterBasedNewVersion
public async Task<string> GetTotalLength(string param1, string param2)
{
    return param1.Length + param2.Length;
}
#endregion

#region ObjectBasedOldVersion
public class TotalLengthRequest
{
    public string Param1 { get; set; }
}

public async Task GetTotalLength(TotalLengthRequest req)
{
    return req.Param1.Length;
}
#endregion

#region ObjectBasedNewVersion
public class TotalLengthRequest
{
    public string Param1 { get; set; }
    public string Param2 { get; set; }
}

public async Task GetTotalLength(TotalLengthRequest req)
{
    var length = req.Param1.Length;
    if (req.Param2 != null)
    {
        length += req.Param2.Length;
    }
    return length;
}
#endregion

#region ClientSideObjectBasedOld
public async Task Broadcast(string message)
{
    await Clients.All.SendAsync("ReceiveMessage", new
    {
        Message = message
    });
}
#endregion

#region ClientSideObjectBasedNew
public async Task Broadcast(string message)
{
    await Clients.All.SendAsync("ReceiveMessage", new
    {
        Sender = Context.User.Identity.Name,
        Message = message
    });
}
#endregion

#region StringlyTyped
public async Task Broadcast(string message)
{
    await Clients.All.SendAsync("ReceiveMessage", new
    {
        Sender = Context.User.Identity.Name,
        Message = message
    });
}
#endregion

#region ClientInterface
public class ChatHubMessage
{
    public string Sender { get; set; }
    public string Message { get; set; }
}

public interface IChatHubClient
{
    Task ReceiveMessage(ChatHubMessage message);
}
#endregion

#region ClientInterfaceHub
public class ChatHub : Hub<IChatHubClient>
{
    public async Task Broadcast(string message)
    {
        await Clients.Add.ReceiveMessage(new ChatHubMessage()
        {
            Sender = Context.User.Identity.Name,
            Message = message
        });
    }
}
#endregion

File: aspnetcore/signalr/api-design/sample/Samples.ts
// <CallWithOneParameter>
connection.invoke("GetTotalLength", "value1");
// </CallWithOneParameter>

// <CallWithObject>
connection.invoke("GetTotalLength", { param1: "value1" });
// </CallWithObject>

// <CallWithObjectNew>
connection.invoke("GetTotalLength", { param1: "value1", param2: "value2" });
// </CallWithObjectNew>

// <OnWithObjectOld>
connection.on("ReceiveMessage", (req) => {
    appendMessageToChatWindow(req.message);
});
// </OnWithObjectOld>

// <OnWithObjectNew>
connection.on("ReceiveMessage", (req) => {
    let message = req.message;
    if (req.sender) {
        message = req.sender + ": " + message;
    }
    appendMessageToChatWindow(message);
});
// </OnWithObjectNew>

// <HubMethodName>
connection.invoke("SendMessage", message);
// </HubMethodName>